docs/design.md

```diff
@@ -0,0 +1,59 @@
+
+## Why Spark Operator?
+
+## Components
+
+The Spark Operator consists of:
+* a `SparkApplication` controller that watches events of creation, updates, and deletion of 
+`SparkApplication` objects and acts on the watch events,
+* a *submission runner* that runs `spark-submit` for submissions received from the controller,
+* a *Spark pod monitor* that watches for Spark pods and sends pod status updates to the controller,
+* a Spark pod [initializer](https://kubernetes.io/docs/admin/extensible-admission-controllers/#initializers)that performs initialization tasks on Spark driver and executor pods based on the annotations on the pods added by the controller,
+* and also a command-line tool named `sparkctl` for working with the operator. 
+
+The following diagram shows how different components interact and work together.
+
+![Architecture Diagram](architecture-diagram.png)
+
+Specifically, a user uses the `sparkctl` (or `kubectl`) to create a `SparkApplication` object. The `SparkApplication` controller receives the object through a watcher from the API server, creates a submission carrying the `spark-submit` arguments, and sends the submission to the *submission runner*. The submission runner submits the application to run and creates the driver pod of the application. Upon starting, the driver pod creates the executor pods. While the application is running, the *Spark pod monitor* watches the pods of the application and sends status updates of the pods back to the controller, which then updates the status of the application accordingly. 
+
+### Controller, Submission Runner, and Spark Pod Monitor
+
+The `SparkApplication` controller, or CRD controller in short, watches events of creation, updates, and deletion of `SparkApplication` objects in any namespaces in a Kubernetes cluster, and acts on the watch events. When it receives a new `SparkApplication` object, it prepares a submission and sends the submission to the submission runner, which actually submits the application to run in the Kubernetes cluster. The submission includes the list of arguments for the `spark-submit` command. The submission runner has a configurable number of workers for submitting applications to run in the cluster. 
```

Comment:
To ensure level-triggered semantics, it should be possible to phrase your CRD controller's behavior in terms like, "For each SparkApplication object that exists, the controller will ..."

The current wording includes things like, "When it receives a new object, it sends a submission," which sounds like edge-triggered logic. In a level-triggered system, you simply observe that an object exists, and you have no idea whether it's the first time you've seen it or not. In the controller design, it's important to define how it will decide whether the observed SparkApplication was already submitted, for example. Or, will it re-submit on every sync, and some downstream component is responsible for deciding whether it was already submitted?

Reply:
> The current wording includes things like, "When it receives a new object, it sends a submission," which sounds like edge-triggered logic. In a level-triggered system, you simply observe that an object exists, and you have no idea whether it's the first time you've seen it or not.

The implementation actually submits whenever the `AddFunc` callback is called, with re-sync disabled though. It's a good point.  

> In the controller design, it's important to define how it will decide whether the observed SparkApplication was already submitted
 
Good point. Will give more thoughts on this.

Reply:
Gave more thoughts on this. The current behavior is the operator submits when a new `SparkApplication` object is added, i.e., when `AddFunc` is called. It also submits if re-submission is necessary, for example if the application should be restarted because of the user-specified `RestartPolicy`. I think the question that remains is how updates to specifications of existing `SparkApplication` objects should be handled. Currently updates are no-ops. 

Stepping back a bit, compared to workloads like deployments, daemonsets, or statefulsets, Spark applications are often batch oriented and are meant to run to completion. So the desire state of a Spark application is simple: running to completion with the driver pod in `succeeded` state eventually. From a Spark perspective, updates to the specification of an application could mean user intentions to re-run the application. A decision needs to be made here regarding how to handle the instance of the application that is currently running when the update happens. However, due to the batch oriented nature, updates not necessarily always mean intentions to re-run. But if this is put into the context of the operator and the declarative nature of the CRD, I think we can safely assume that an update to the specification of a `SparkApplication` object carries the user intention to re-run the application with the updated spec.

To track thoughts and work on handling updates, I created https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/issues/83.